File: 백준/Silver/9012. 괄호/괄호.c
#include <stdio.h>
#include <stdbool.h>

int main(void){
    char str[51];
    int T;
    scanf("%d", &T);
    while(T--){
        scanf("%s", str);
        int left = 0;
        bool fail = false;
        for(int i = 0; str[i]; i++){
            if(str[i] == '(') left++;
            else{
                if(!left){
                    fail = true;
                    break;
                }else{
                    left--;
                }
            }
        }
        if(left) fail = true;
        printf("%s\n", fail ? "NO" : "YES");
    }
}

File: 백준/Silver/9012. 괄호/괄호.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) throws Exception {
        Scanner scanner = new Scanner(System.in);
        int T = scanner.nextInt();
        scanner.nextLine();

        String str;
        for(int i = 0; i < T; i++){
            str = scanner.nextLine();
            int len = str.length();
            int left = 0;
            boolean fail = false;

            for(int k = 0; k < len; k++){
                if(str.charAt(k) == '(') left++;
                else{
                    if(left == 0) {
                        fail = true;
                        break;
                    }else left--;
                }
            }
            if(left > 0) fail = true;
            System.out.println(fail ? "NO" : "YES");
        }
    }
}
